src/Compilers/Core/Portable/DocumentationCommentId.cs

```diff
@@ -301,64 +312,113 @@ private static string DecodePropertyName(string name, string language)
         {
             private readonly StringBuilder _builder;
-            private readonly Generator _generator;
+            private readonly DeclarationGenerator _generator;
 
-            public DeclarationGenerator(StringBuilder builder)
+            public PrefixAndDeclarationGenerator(StringBuilder builder)
             {
                 _builder = builder;
-                _generator = new Generator(builder);
+                _generator = new DeclarationGenerator(builder);
             }
 
-            public override void DefaultVisit(ISymbol symbol)
-            {
-                throw new InvalidOperationException("Cannot generated a documentation comment id for symbol.");
-            }
+            /// <summary>
+            /// If we hit anything we don't know about, indicate failure.
+            /// </summary>
+            public override bool DefaultVisit(ISymbol symbol)
+                => false;
 
-            public override void VisitEvent(IEventSymbol symbol)
+            public override bool VisitEvent(IEventSymbol symbol)
             {
                 _builder.Append("E:");
-                _generator.Visit(symbol);
+                return _generator.Visit(symbol);
             }
 
-            public override void VisitField(IFieldSymbol symbol)
+            public override bool VisitField(IFieldSymbol symbol)
             {
                 _builder.Append("F:");
-                _generator.Visit(symbol);
+                return _generator.Visit(symbol);
             }
 
-            public override void VisitProperty(IPropertySymbol symbol)
+            public override bool VisitProperty(IPropertySymbol symbol)
             {
                 _builder.Append("P:");
-                _generator.Visit(symbol);
+                return _generator.Visit(symbol);
             }
 
-            public override void VisitMethod(IMethodSymbol symbol)
+            public override bool VisitMethod(IMethodSymbol symbol)
             {
                 _builder.Append("M:");
-                _generator.Visit(symbol);
+                return _generator.Visit(symbol);
             }
 
-            public override void VisitNamespace(INamespaceSymbol symbol)
+            public override bool VisitNamespace(INamespaceSymbol symbol)
             {
                 _builder.Append("N:");
-                _generator.Visit(symbol);
+                AppendNamespace(symbol, _builder);
+                return true;
             }
 
-            public override void VisitNamedType(INamedTypeSymbol symbol)
+            public override bool VisitNamedType(INamedTypeSymbol symbol)
             {
                 _builder.Append("T:");
-                _generator.Visit(symbol);
+                _generator.VisitNamedType(symbol);
+                return true;
+            }
+
+            private static void AppendNamespace(INamespaceSymbol symbol, StringBuilder builder)
```

Comment:
moved the visit methods into these explicit namespace/named-type helpers.  it makes it much clearer that there are no failure paths here.